### ISSUE TITLE
Impossible to satisfy const-correctness

### ISSUE THREAD

**Issue by Y-Less** [pawn-lang/compiler#371]
### Issue description:

When you have two states for a function, one of which modifies an array, the other of which doesn't, the compiler will complain about the const-correctness of the argument regardless of what you do.

### Minimal complete verifiable example (MCVE):

```pawn
MyFunc(str[], &val) <my_state : a>
{
	val = str[0];
}

MyFunc(str[], &val) <my_state : b>
{
	str[0] = val;
}

main()
{
	state my_state : a;
	new b, arr[10];
	MyFunc(arr, b);
}
```

`(1) : warning 214: possibly a "const" array argument was intended: "str"`

```pawn
MyFunc(const str[], &val) <my_state : a>
{
	val = str[0];
}

MyFunc(str[], &val) <my_state : b>
{
	str[0] = val;
}

main()
{
	state my_state : a;
	new b, arr[10];
	MyFunc(arr, b);
}
```

`(6) : error 025: function heading differs from prototype`

```pawn
MyFunc(const str[], &val) <my_state : a>
{
	val = str[0];
}

MyFunc(const str[], &val) <my_state : b>
{
	str[0] = val;
}

main()
{
	state my_state : a;
	new b, arr[10];
	MyFunc(arr, b);
}
```

`(8) : error 022: must be lvalue (non-constant)`
`(8) : warning 215: expression has no effect`

```
```

I did discover that you can shut the warning up with this:

```pawn
MyFunc(str[], &val) <my_state : a>
{
	#pragma unused str
	val = str[0];
}

MyFunc(str[], &val) <my_state : b>
{
	str[0] = val;
}

main()
{
	state my_state : a;
	new b, arr[10];
	MyFunc(arr, b);
}
```

I don't know if the fact that that works is a bug in itself...
### Workspace Information:

* Compiler version:
* Command line arguments provided (or sampctl version):
* Operating System:

**Comment by Y-Less** [comment]
Clearly, since `MyFunc` compiles to a single function, then `str` is at some point modified, so there should be no warning in any state.

**Comment by YashasSamaga** [comment]
`#pragma unused` works by marking the symbol as `uREAD` and `uWRITTEN` irrespective of whether it was actually read or written. Hence, `const` qualifier is not suggested as the symbol is considered to be modified.

**Comment by YashasSamaga** [comment]
```
MyFunc(&val) <my_state : a>
{
	val = 0;
	//val = str[0];
}

MyFunc(&val) <my_state : b>
{
	//str[0] = val;
}

main()
{
	state my_state : a;
	new b;
	MyFunc(b);
}
```

causes the compiler to complain that `val` is unused.

Is the compiler correct in this case?

**Comment by Y-Less** [comment]
That's a very good point about unused variables. I think the better solution could just be the pragma, since that is the solution in the unused variable case as well.

**Comment by YashasSamaga** [comment]
Isn't this actually a bug? The warning is clearly invalid and shouldn't have been triggered at all.

**Comment by Daniel-Cortez** [comment]
Please close this (fixed in #372).
